<launch>
  <node name="object_singulation_perception" pkg="tabletop_pushing" type="object_singulation_node" respawn="false" output="screen">
    <remap from="color_image_topic" to="/kinect_head/rgb/image_color"/>
    <remap from="depth_image_topic" to="/kinect_head/depth/image"/>
    <remap from="point_cloud_topic" to="/kinect_head/depth/points"/>

    <param name="use_displays" value="true" type="bool"/>
    <param name="display_wait_ms" value="3" type="int"/>
    <param name="autostart_pcl_segmentation" value="false" type="bool"/>
    <param name="use_guided_pushes" value="true" type="bool"/>
    <param name="write_to_disk" value="true" type="bool"/>
    <param name="write_input_to_disk" value="true" type="bool"/>

    <param name="crop_min_x" value="22" type="int"/>
    <param name="crop_max_x" value="593" type="int"/>
    <param name="crop_min_y" value="32" type="int"/>
    <param name="crop_max_y" value="469" type="int"/>

    <param name="workspace_frame" value="/torso_lift_link" type="string"/>
    <param name="min_workspace_x" value="-1.0" type="double"/>
    <param name="max_workspace_x" value="1.75" type="double"/>
    <param name="min_workspace_y" value="-1.2" type="double"/>
    <param name="max_workspace_y" value="1.2" type="double"/>
    <param name="min_workspace_z" value="-0.8" type="double"/>
    <param name="max_workspace_z" value="0.6" type="double"/>
    <param name="below_table_z" value="0.1" type="double"/>

    <param name="num_downsamples" value="1" type="int"/>

    <param name="pcl_cluster_tolerance" value="0.03" type="double"/>
    <param name="pcl_min_cluster_size" value="200" type="int"/>
    <param name="pcl_max_cluster_size" value="25000" type="int"/>
    <param name="pcl_voxel_downsample_res" value="0.005" type="double"/>
    <param name="pcl_cloud_intersect_thresh" value="0.0025" type="double"/>
    <param name="pcl_concave_hull_alpha" value="0.1" type="double"/>
    <param name="use_pcl_voxel_downsample" value="false" type="bool"/>
    <param name="table_ransac_thresh" value="0.015" type="double"/>
    <param name="table_ransac_angle_thresh" value="5.0" type="double"/>
    <param name="pcl_difference_thresh" value="0.0001" type="double"/>

    <param name="min_table_z" value="-1.0" type="double"/>
    <param name="max_table_z" value="1.0" type="double"/>
    <param name="min_pushing_y" value="-0.4" type="double"/>
    <param name="max_pushing_y" value="0.4" type="double"/>
    <param name="min_pushing_x" value="0.35" type="double"/>
    <param name="max_pushing_x" value="0.85" type="double"/>
    <!-- <param name="min_pushing_angle" value="-1.885" type="double"/> -->
    <!-- <param name="max_pushing_angle" value="1.885" type="double"/> -->
    <param name="min_push_dist" value="0.07" type="double"/>
    <param name="max_push_dist" value="0.3" type="double"/>
    <param name="push_dist_inflation" value="0.02" type="double"/>

    <param name="threshold_edges" value="true" type="bool"/>
    <param name="edge_weight_thresh" value="0.1" type="double"/>
    <param name="depth_edge_weight_thresh" value="0.05" type="double"/>
    <param name="depth_edge_weight" value="0.75" type="double"/>

    <param name="boundary_ransac_thresh" value="0.005" type="double"/>
    <param name="min_edge_length" value="4" type="int"/>
    <param name="os_min_cluster_size" value="200" type="int"/>
    <param name="os_min_boundary_length" value="10" type="int"/>
    <param name="os_hist_bin_width" value="5" type="int"/>
    <param name="os_hist_bin_height" value="15" type="int"/>
    <param name="push_collision_thresh" value="0.03" type="double"/>
    <param name="start_collision_thresh" value="0.05" type="double"/>
    <param name="num_angle_bins" value="4" type="int"/>
    <!-- <param name="img_output_path" value="/home/thermans/data/new/" type="string"/> -->
    <param name="img_output_path" value="/u/thermans/data/new/" type="string"/>
    <param name="moved_count_thresh" value="3" type="int"/>
    <param name="force_remain_singulated" value="false" type="bool"/>

    <param name="bad_icp_score_limit" value="0.0001" type="double"/>
    <param name="icp_max_iters" value="1000" type="int"/>
    <param name="icp_transform_eps" value="0" type="double"/>
    <param name="icp_max_cor_dist" value="1.0" type="double"/>
    <param name="icp_ransac_thresh" value="0.5" type="double"/>

    <param name="per_object_rand_push_count" value="1" type="int"/>
    <param name="use_unguided_icp" value="true" type="bool"/>
    <param name="push_guess_limit" value="16" type="int"/>
  </node>
</launch>
